File: agent/src/home_iot/routines/night.py
"""Deterministic night routine action planner.

This is intentionally not an LLM module. It converts the current context into a
small list of candidate actions. Later layers can decide whether to speak them,
call HA services, or ask an LLM to phrase them.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any


def plan_night_routine(context: dict[str, Any], *, now: datetime) -> list[dict[str, Any]]:
    if not _is_evening_window(now):
        return []

    routines = _dict(context.get("routines"))
    if routines.get("night_briefing_done") == "on":
        return []

    computer = _dict(context.get("computer"))
    if _should_defer_for_focus(computer):
        return []

    actions: list[dict[str, Any]] = []
    supplements = _dict(context.get("supplements"))
    if supplements.get("magnesium") == "missing":
        actions.append(
            {
                "id": "take_magnesium",
                "priority": "medium",
                "message": "마그네슘 아직 안 챙기셨어요.",
            }
        )

    environment = _dict(context.get("environment"))
    bedroom_co2 = _number(environment.get("bedroom_co2"))
    if bedroom_co2 is not None and bedroom_co2 >= 900:
        actions.append(
            {
                "id": "ventilate_bedroom",
                "priority": "medium",
                "message": f"침실 CO2가 {int(bedroom_co2)}ppm이라 10분 환기 추천드려요.",
            }
        )

    if actions:
        actions.append(
            {
                "id": "dim_lights_for_bedtime",
                "priority": "low",
                "message": "취침 준비를 위해 침실 조명을 낮춰둘게요.",
                "ha_service": {
                    "domain": "light",
                    "service": "turn_on",
                    "target": {"entity_id": "light.bedroom"},
                    "data": {"brightness_pct": 30},
                },
            }
        )

    return actions


def _is_evening_window(now: datetime) -> bool:
    # Keep MVP simple: 21:00–23:59 local/runtime time.
    return 21 <= now.hour <= 23


def _should_defer_for_focus(computer: dict[str, Any]) -> bool:
    if computer.get("in_meeting") == "on":
        return True
    focus_minutes = _number(computer.get("focus_block_minutes"), 0) or 0
    if computer.get("current_mode") == "coding" and focus_minutes >= 45:
        return True
    return False


def _dict(value: Any) -> dict[str, Any]:
    return value if isinstance(value, dict) else {}


def _number(value: Any, default: float | None = None) -> float | None:
    if isinstance(value, int | float):
        return float(value)
    if isinstance(value, str):
        try:
            return float(value)
        except ValueError:
            return default
    return default

File: personal-home-os/src/routines/night.rs
use crate::context::{ActionSafety, HaServiceCall, NowContext, RecommendedAction};

pub fn plan_night_routine(context: &NowContext) -> Vec<RecommendedAction> {
    if context.routines.night_briefing_done || context.computer.in_meeting || is_deep_focus(context)
    {
        return Vec::new();
    }

    let mut actions = Vec::new();

    if context.supplements.magnesium == "missing" {
        actions.push(RecommendedAction {
            id: "take_magnesium".to_string(),
            title: "마그네슘 챙기기".to_string(),
            priority: "medium".to_string(),
            message: "마그네슘은 아직 안 챙기셨어요.".to_string(),
            reason: vec!["오늘 magnesium taken 이벤트 없음".to_string()],
            safety: ActionSafety {
                auto_executable: false,
                requires_approval: false,
            },
            ha_service: None,
        });
    }

    if context.environment.bedroom_co2 >= 1000 {
        actions.push(RecommendedAction {
            id: "ventilate_bedroom".to_string(),
            title: "침실 10분 환기".to_string(),
            priority: "medium".to_string(),
            message: "침실 CO2가 높아서 10분 환기를 추천해요.".to_string(),
            reason: vec![
                format!("bedroom_co2={}ppm", context.environment.bedroom_co2),
                "취침 시간대".to_string(),
                "회의 중 아님".to_string(),
            ],
            safety: ActionSafety {
                auto_executable: false,
                requires_approval: true,
            },
            ha_service: None,
        });
    }

    if context.environment.bedroom_light_brightness > 30 {
        actions.push(RecommendedAction {
            id: "dim_lights_for_bedtime".to_string(),
            title: "침실 조명 낮추기".to_string(),
            priority: "low".to_string(),
            message: "취침 준비를 위해 침실 조명을 낮춰둘 수 있어요.".to_string(),
            reason: vec![
                format!(
                    "bedroom_light_brightness={}pct",
                    context.environment.bedroom_light_brightness
                ),
                "취침 루틴".to_string(),
            ],
            safety: ActionSafety {
                auto_executable: true,
                requires_approval: false,
            },
            ha_service: Some(HaServiceCall {
                domain: "light".to_string(),
                service: "turn_on".to_string(),
                entity_id: "light.bedroom".to_string(),
                brightness_pct: Some(30),
            }),
        });
    }

    actions
}

fn is_deep_focus(context: &NowContext) -> bool {
    context.computer.current_mode == "coding" && context.computer.focus_block_minutes >= 45
}
